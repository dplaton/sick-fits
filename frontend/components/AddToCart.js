import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import Error from './ErrorMessage';

const ADD_TO_CART_MUTATION = gql`
    mutation addToCart($id: ID!) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`;

class AddToCart extends React.Component {
    render() {
        const { id } = this.props;

        return (
            <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ id }}>
                {(addToCart, {error})=>(
                    <button onClick={addToCart}>Add to cart ðŸ›’</button>
                )}
            </Mutation>
        );
    }
}

export default AddToCart;
