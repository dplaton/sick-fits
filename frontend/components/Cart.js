import React from 'react';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';

const Cart = () => {
    return(
        <CartStyles open>
            <header>
                <CloseButton title="close">&times;</CloseButton>
                <Supreme>Your shopping cart</Supreme>
                <p>You have __ items in cart</p>
            </header>
            <footer>
                <p>$10.10</p>
                <SickButton>Checkout</SickButton>
            </footer>
        </CartStyles>
    )
}

export default Cart;