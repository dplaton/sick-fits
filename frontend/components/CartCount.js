import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition } from "react-transition-group";

/*
 * We customize the classes of the component so we have a flipping effect   
*/
const AnimationStyles = styled.span`
    position: relative;
    .count {
        display: block;
        position: relative;
        transition: all .4s;
        backface-visibility: hidden;
    }
    .count-enter {
        transform: scale(4) rotateX(0.5turn);
    }
    .count-enter-active {
        transform: rotateX(0);
    }
    .count-exit {
        top:0;
        position:absolute;
        transform: rotateX(0);
    }
    .count-exit-active {
        transform: scale(4) rotateX(0.5turn);
    }
`;

const Dot = styled.div`
    background: ${props => props.theme.red};
    border-radius: 50%;
    color: white;
    font-weight: 100;
    font-feature-settings: "tnum";
    font-variant-numeric: tabular-nums;
    line-height: 2rem;
    min-width: 3rem;
    margin-left: 1rem;
    padding: 0.5rem;
`;

/*
 * Using the TransitionGroup and the CSSTransition allows us to add a nice animation to the "cart count" badge
 * What React is doing is basically unmounting a component with count = x and mounting another one with count=x+1. 
 * In this proces React exposes a bunch of CSS classes so we can style the components (both the unmounted one and the mounted one)
 * 
 */
const CartCount = ({ count }) => (
    <AnimationStyles>
        <TransitionGroup>
            <CSSTransition
                unmountOnExit
                className="count"
                classNames="count"
                timeout={{ enter: 400, exit: 400 }}
                key={count}
            >
                <Dot>{count}</Dot>
            </CSSTransition>
        </TransitionGroup>
    </AnimationStyles>
);

export default CartCount;
