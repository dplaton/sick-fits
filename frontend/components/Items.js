import React, {Component} from 'react'
import gql from 'graphql-tag'
import {Query} from 'react-apollo'

const ALL_ITEMS_QUERY = gql `
    query ALL_ITEMS_QUERY {
        items {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

class Items extends Component {
    render() {
        return (
            <div>
                <Query query={ALL_ITEMS_QUERY}>
                    {
                        ({data, error, loading}) => {
                            if (loading) return <p>Loading...</p>
                            if (error) return <p>Error! {error}</p>

                            return <p>Hey, I found {data.items.length}</p>;
                        }
                    } 
                </Query>
            </div>
        )
    }
}

export default Items