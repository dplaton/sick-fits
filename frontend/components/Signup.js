import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import Form from "./styles/Form";
import Error from "./ErrorMessage";

const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION(
        $email: String!
        $name: String!
        $password: String!
    ) {
        signUp(email: $email, name: $name, password: $password) {
            id
            email
            name
        }
    }
`;

class Signup extends Component {
    state = {
        name: "",
        email: "",
        password: ""
    };

    saveToState = event => {
        this.setState({ [event.target.name]: event.target.value });
    };

    render() {
        return (
            <Mutation mutation={SIGNUP_MUTATION} variables={this.state}>
                {(signUp, { error, loading }) => {
                    return (
                        <Form
                            method="POST"
                            onSubmit={async e => {
                                e.preventDefault();
                                await signUp();
                                this.setState({name:'', email:'', password:''});
                            }}
                        >
                            <fieldset disabled={loading} aria-busy={loading}>
                                <Error error={error} />
                                <h2>Sign up for an account</h2>
                                <label htmlFor="email">
                                    Email
                                    <input
                                        type="email"
                                        name="email"
                                        placeholder="Your email"
                                        value={this.state.email}
                                        onChange={this.saveToState}
                                    />
                                </label>
                                <label htmlFor="password">
                                    Password
                                    <input
                                        type="password"
                                        name="password"
                                        placeholder="Your password"
                                        value={this.state.password}
                                        onChange={this.saveToState}
                                    />
                                </label>
                                <label htmlFor="name">
                                    Name
                                    <input
                                        type="text"
                                        name="name"
                                        placeholder="Your name"
                                        value={this.state.name}
                                        onChange={this.saveToState}
                                    />
                                </label>
                                <button type="submit">Create account</button>
                            </fieldset>
                        </Form>
                    );
                }}
            </Mutation>
        );
    }
}

export default Signup;
