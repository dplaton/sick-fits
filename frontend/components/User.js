import { Query } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";

const CURRENT_USER_QUERY = gql`
    query {
        me {
            id
            name
            email
            permissions
            cart {
                id
                quantity
                item {
                    id
                    title
                    price
                    image
                    description
                }
            }
        }
    }
`;

// this is how we build our own render prop
const User = props => (
    // the children of this component must be functions which receive the 'payload' as argument
    <Query {...props} query={CURRENT_USER_QUERY}>
        {payload => props.children(payload)}
    </Query>
);

User.propTypes = {
    children: PropTypes.func.isRequired
};

export default User;
export { CURRENT_USER_QUERY };
