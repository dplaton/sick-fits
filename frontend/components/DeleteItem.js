import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';

import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
        }
    }
`;

class DeleteItem extends Component {
    update = (cache, payload) => {
        // manually update the cache on the client instead of re-fetching everything

        // 1. Read the items from the cache
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });

        // 2. Filter the deleted item
        data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
    };

    render() {
        return (
            // relevant docs: https://www.apollographql.com/docs/react/essentials/mutations.html
            // the "udpdate" prop specifies a function to be used as a 'callback' after the mutation was executed
            <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ id: this.props.id }} update={this.update}>
                {(deleteItem, { error }) => (
                    <button
                        onClick={() => {
                            if (confirm('Are you sure you want to delete this item?')) {
                                deleteItem()
                                    .catch(err => { alert(err.message) });
                            }
                        }}>
                        {this.props.children}
                    </button>
                )}
            </Mutation>
        );
    }
}

export default DeleteItem;
