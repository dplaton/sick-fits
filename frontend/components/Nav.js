import Link from "next/link";
import NavStyles from "./styles/NavStyles";
import User from './User'

const Nav = () => (
    <NavStyles>
        <User>
            {({data: {me}}) => {
                // two level destructuring, like payload.data.me
                if (me) {
                    return <p>{me.name}</p>
                } else {
                    return <p>Nobody</p>
                }
            }}
        </User>
        <Link href="/items">
            <a>Shop</a>
        </Link>
        <Link href="/sell">
            <a>Sell</a>
        </Link>
        <Link href="/signup">
            <a>Signup</a>
        </Link>
        <Link href="/orders">
            <a>Orders</a>
        </Link>
        <Link href="/me">
            <a>Account</a>
        </Link>
    </NavStyles>
);

export default Nav;
