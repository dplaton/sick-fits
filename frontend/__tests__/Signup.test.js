import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";

import Signup, { SIGNUP_MUTATION } from "../components/Signup";
import { fakeUser } from "../lib/testUtils";
import { CURRENT_USER_QUERY } from "../components/User";

const me = fakeUser();
const mocks = [
    // signup mock query
    {
        request: {
            query: SIGNUP_MUTATION,
            variables: {
                name: me.name,
                email: me.email,
                password: "done"
            }
        },
        result: {
            data: {
                signUp: {
                    __typename: "User",
                    id: "abc123",
                    email: me.email,
                    name: me.name
                }
            }
        }
    },
    // user query mock
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me
            }
        }
    }
];

function type(wrapper, name, value) {
    wrapper.find(`input[name="${name}"]`).simulate("change", {
        target: {
            name,
            value
        }
    });
}

describe("<Signup/>", () => {
    it("renders and matches snapshot", () => {
        const wrapper = mount(
            <MockedProvider>
                <Signup />
            </MockedProvider>
        );

        expect(toJSON(wrapper.find("form"))).toMatchSnapshot();
    });

    it("executes the mutation properly", async () => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    {client => {
                        apolloClient = client;
                        return <Signup />;
                    }}
                </ApolloConsumer>
            </MockedProvider>
        );

        await wait();
        wrapper.update();

        type(wrapper, "email", me.email);
        type(wrapper, "password", "done");
        type(wrapper, "name", me.name);

        wrapper.find("form").simulate("submit");
        await wait();

        const result = await apolloClient.query({ query: CURRENT_USER_QUERY });
        expect(result.data.me).toMatchObject(me);
    });
});
