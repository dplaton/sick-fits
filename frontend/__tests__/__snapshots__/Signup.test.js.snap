// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  method="POST"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <DisplayError
      error={Object {}}
    />
    <h2>
      Sign up for an account
    </h2>
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Your email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Your password"
        type="password"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="Your name"
        type="text"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Create account
    </button>
  </fieldset>
</form>
`;
